content/blog/2019-01-07-we-mastered-the-art-of-programming.md

```diff
@@ -0,0 +1,63 @@
+
+What a great modification done by a great programmer.
+
+## Wait.
+
+This is the moment the epiphany came. I was not the one have been doing a great job: my teammate was. He handed me some code that was so clean that anyone could quickly modify anything without unexpected side effect. Everything was self-explanatory, no comments lying around, no arbitrary numbers. The code had the perfect vertical and horizontal length and my solarized theme made the code look like a work of art. I experienced, what I thought at that time was, the holy grail of good code.
+
+## Different solutions.
+
+First of all, there were things my teammate knew that i didn't know that i didn't know. Things like data pipeline, [design patterns](https://sourcemaking.com/design_patterns) and [principles](https://en.wikipedia.org/wiki/SOLID), unit testing, [clean code](https://en.wikipedia.org/wiki/Worship), software architecture, etc. These elements are the normal distribution for the actuary, the mean and variance for the statistician. The ground on which everything is build. Clearly my hacking and school background never prepared me for this.
+
+My teammate had an holistic view of the problem. I was too busy *making things work* while my teammate was busy figuring *the best way* to make it work. This subtle difference is crucial. A good solution will make something work right now. The best solution makes something work right now AND consider the future user. Even if it means taking some extra time to step back and rebuild in the right direction.
+
+Clearly no solution could (or should) have initially planted all the seeds for the craziness that came after the first request. But my stubbornness to produce a quick solution building on my first design without reconsidering the paradigm shift in the problem doomed me in the long run.
+
+## Clean slate.
+
+I enrolled in a software engineering certificate to rebuild a more stable foundation. This allowed me to open a huge Pandora’s box that uncover thought-provoking and mind breaking ideas and concepts. I believe I’m a better programmer than I was and I believe I have the tools to be a better one tomorrow.
+
+This articles is the first of many that will allow me to share those tools I found in that Pandora's box. These articles will be half philosophical and half technical. They will be centered around good practices in software engineering, while sharing some of my experience coping with (really) crappy code base. Crappy code base that have been written by me, of course.
```

Comment:
this article

Reply:
first of many in which I will share these

Reply:
base that had ou has been

Reply:
xie xie fixed